src/coreclr/vm/mlinfo.cpp

```diff
@@ -1056,6 +1056,78 @@ OleColorMarshalingInfo *EEMarshalingData::GetOleColorMarshalingInfo()
 }
 #endif // FEATURE_COMINTEROP
 
+namespace
+{
+    MarshalInfo::MarshalType GetDisabledMarshallingMarshalerType(
+        Module* pModule,
+        SigPointer sig,
+        const SigTypeContext * pTypeContext,
+        MethodTable** pMTOut,
+        UINT* errorResIDOut)
+    {
+        switch (sig.PeekElemTypeNormalized(pModule, pTypeContext))
+        {
+        case ELEMENT_TYPE_BOOLEAN:
+        case ELEMENT_TYPE_U1:
+            return MarshalInfo::MARSHAL_TYPE_GENERIC_U1;
+        case ELEMENT_TYPE_I1:
+            return MarshalInfo::MARSHAL_TYPE_GENERIC_1;
+        case ELEMENT_TYPE_CHAR:
+        case ELEMENT_TYPE_U2:
+            return MarshalInfo::MARSHAL_TYPE_GENERIC_U2;
+        case ELEMENT_TYPE_I2:
+            return MarshalInfo::MARSHAL_TYPE_GENERIC_2;
+        case ELEMENT_TYPE_U4:
+            return MarshalInfo::MARSHAL_TYPE_GENERIC_U4;
+        case ELEMENT_TYPE_I4:
+            return MarshalInfo::MARSHAL_TYPE_GENERIC_4;
+        case ELEMENT_TYPE_U8:
+        case ELEMENT_TYPE_I8:
+            return MarshalInfo::MARSHAL_TYPE_GENERIC_8;
+#ifdef TARGET_64BIT
+        case ELEMENT_TYPE_U:
+        case ELEMENT_TYPE_PTR:
+        case ELEMENT_TYPE_FNPTR:
+        case ELEMENT_TYPE_I:
+            return MarshalInfo::MARSHAL_TYPE_GENERIC_8;
+#else
+        case ELEMENT_TYPE_U:
+            return MarshalInfo::MARSHAL_TYPE_GENERIC_U4;
+        case ELEMENT_TYPE_PTR:
+        case ELEMENT_TYPE_FNPTR:
+        case ELEMENT_TYPE_I:
+            return MarshalInfo::MARSHAL_TYPE_GENERIC_4;
+#endif
+        case ELEMENT_TYPE_R4:
+            return MarshalInfo::MARSHAL_TYPE_FLOAT;
+        case ELEMENT_TYPE_R8:
+            return MarshalInfo::MARSHAL_TYPE_DOUBLE;
+        case ELEMENT_TYPE_VAR:
+        case ELEMENT_TYPE_VALUETYPE:
+        {
+            TypeHandle sigTH = sig.GetTypeHandleThrowing(pModule, pTypeContext);
+            MethodTable* pMT = sigTH.GetMethodTable();
+
+            if (!pMT->IsValueType() || pMT->ContainsPointers())
```

Comment:
Why would the element type be `ELEMENT_TYPE_VALUETYPE`, but `IsValueType` be false?

Reply:
This is also under the `ELEMENT_TYPE_VAR` case expression, so we will also hit this for generics. Now, I don't think we actually have any possible path where we can get an `ELEMENT_TYPE_VAR` into the `MarshalInfo` constructor since we removed the WinRT support in the runtime and we don't support P/Invokes or interop delegates in generic classes, but I didn't want to assert that as part of this change. I'd rather do that as a follow up clean-up.